LTP/ima: grub setup

pevik opened the pull request:
The goal is to boot kernel with parameters for IMA while keep default parameters the rest of LTP tests.
This is required by ima_measurements test which fails (https://openqa.suse.de/tests/1711511).

**ORIGINAL grub**
- SLE15
http://quasar.suse.cz/tests/176#step/install_ltp/1
- openSUSE
http://quasar.suse.cz/tests/177#step/install_ltp/1
** NEW grub (2 more enties)**
- SLE15
http://quasar.suse.cz/tests/182#step/boot_ltp/1
- openSUSE
http://quasar.suse.cz/tests/181#step/boot_ltp/1

Disadvantages: it uses the same needles. Maybe it'd be good to at least check in needles / in grub output there are more grub entries. I didn't implement it because I use `assert_script_run` for `sed` and `grub2-mkconfig`. Maybe adding counting number of entries reported by `grub2-mkconfig` would be enough.

**IMA tests**
- choose correct kernel SLE15
http://quasar.suse.cz/tests/189#step/boot_ltp/2
- choose correct kernel openSUSE
http://quasar.suse.cz/tests/188#step/boot_ltp/2
-  /proc/cmdline is with ima_policy=tcb (SLE15)
http://quasar.suse.cz/tests/189/file/boot_ltp-ver_linux_before.txt
BOOT_IMAGE=/boot/vmlinuz-4.12.14-23-default root=/dev/vda2 ima_policy=tcb video=1024x768-16 plymouth.ignore-serial-consoles console=ttyS0 console=tty resume=/dev/disk/by-path/pci-0000:00:06.0-part3 splash=silent showopts
-  /proc/cmdline is with ima_policy=tcb (openSUSE)
http://quasar.suse.cz/tests/188/file/boot_ltp-ver_linux_before.txt
BOOT_IMAGE=/boot/vmlinuz-4.16.12-1-default root=/dev/vda2 ima_policy=tcb video=1024x768-16 plymouth.ignore-serial-consoles console=ttyS0 console=tty resume=/dev/disk/by-path/pci-0000:00:06.0-part3 splash=silent showopts

**non IMA tests**
- choose default kernel SLE15
http://quasar.suse.cz/tests/190#step/boot_ltp/1
- choose default kernel openSUSE
http://quasar.suse.cz/tests/191#step/boot_ltp/1
- /proc/cmdline is without ima_policy=tcb (SLE)
http://quasar.suse.cz/tests/190/file/boot_ltp-ver_linux_before.txt
BOOT_IMAGE=/boot/vmlinuz-4.12.14-23-default root=/dev/vda2 video=1024x768-16 plymouth.ignore-serial-consoles console=ttyS0 console=tty resume=/dev/disk/by-path/pci-0000:00:06.0-part3 splash=silent showopts
- /proc/cmdline is without ima_policy=tcb (openSUSE)
http://quasar.suse.cz/tests/191/file/boot_ltp-ver_linux_before.txt
BOOT_IMAGE=/boot/vmlinuz-4.16.12-1-default root=/dev/vda2 video=1024x768-16 plymouth.ignore-serial-consoles console=ttyS0 console=tty resume=/dev/disk/by-path/pci-0000:00:06.0-part3 splash=silent showopts

Implements: poo#35637

pevik commented:
@richiejp @mimi1vx WIP is for not only for not wanting to break kernel, but also that some parameters obvisously didn't propagate (IMA kernel lacks `video=1024x768-16 plymouth.ignore-serial-consoles console=ttyS0 console=tty resume=/dev/disk/by-path/pci-0000:00:06.0-part3 splash=silent showopts
`, I should fix that)

richiejp commented on tests/kernel/install_ltp.pm:
```diff
@@ -165,10 +166,29 @@ sub install_from_repo {
     assert_script_run q(find ${LTPROOT:-/opt/ltp}/testcases/bin/openposix/conformance/interfaces/ -name '*.run-test' > ~/openposix-test-list-) . $tag;
 }
 
+sub add_custom_grub_entries {
```
Maybe this should go inside the util or boot library modules?

richiejp commented on tests/kernel/install_ltp.pm:
```diff
@@ -165,10 +166,29 @@ sub install_from_repo {
+    my $self       = shift;
+    my $inst_ltp   = get_var 'INSTALL_LTP';
+    my $tag        = get_var('LTP_RUNTEST_TAG') || get_var('VERSION') . '-' . get_var('BUILD');
+    my $grub_param = get_var('LTP_GRUB_PARAM') || 'ima_policy=tcb';
```
maybe just call this GRUB_PARAM and add a new test module which adds boot options. I imagine this might be useful for a lot of tests besides LTP

pevik replied:
Well, before I created grub_utils and send pull request https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/5030, but it wasn't really useful. This is different code, maybe reusable for somebody => ok, I'll try to add it into `bootloader_setup.pm` (hope it wont be wasted time again).

mimi1vx commented:
@pevik I have bad news fro you, we started testing also SLE11 kernels using openQA:D

pevik commented:
@mimi1vx good to know. Hope SLE11 is not affected by this (I didn't have to change needles for this).

pevik commented:
FYI: Pushed new version, not tested yet.

pevik commented:
@richiejp @mimi1vx I tested new version. Works well, but safer to merge after SLE15 gets GM (or at least after GMC testing).
Changelog:
Adding methods into `lib/bootloader_setup.pm`
Changing boot to default IMA (not going to advanced).

richiejp commented:
LGTM

pevik commented:
@richiejp @mimi1vx @SeroSun FYI: I added `GRUB_PARAM=ima_policy=tcb` to `install_ltp*` in both https://openqa.suse.de/admin/test_suites and https://openqa.opensuse.org/admin/test_suites . It doesn't affect anything until this is merged.
The only thing which haven't been tested is `install_ltp+sle+Server-DVD-KOTD`, `install_ltp+sle+15+Installer-DVD-m32` and `install_ltp+sle+Server-DVD+CVE`
I added it to all jobs, for consistency, although the latter 2 (m32 and Richard's CVE) obviously don't need that.

pevik commented:
@mimi1vx \+ SLE 11 hasn't been tested, as there are not any jobs to clone in  https://openqa.suse.de/group_overview/161 https://openqa.suse.de/group_overview/160

okurz commented on tests/kernel/boot_ltp.pm:
```diff
@@ -14,16 +14,23 @@ use 5.018;
-    # During install_ltp, the second boot may take longer than usual.
-    $self->wait_boot(ready_time => 500);
+    if ($is_ima) {
+        $self->boot_grub_item(2);
```
That 2 reads quite magic to me. Can this be improved to be more explicit?

okurz commented on tests/kernel/install_ltp.pm:
```diff
@@ -271,5 +272,10 @@ will be used.
 
 Overrides the official LTP GitHub repository URL.
 
+=head2 GRUB_PARAM
```
I wonder is this documentation somewhere rendered? Any doc server or something?

pevik replied:
I have no idea.

foursixnine replied:
I really doubt this is published, but @okurz generating and publishing in http://open.qa would be a nice idea (Also for the rest of the test distribution... but I don't know if it's really needed, but I do see a benefit).

okurz commented on lib/bootloader_setup.pm:
```diff
@@ -56,6 +58,91 @@ sub stop_grub_timeout {
+    for my $i (1 .. $menu1) {
+        send_key "down";
+    }
+    sleep 1;
```
Please avoid sleeps. Better wrap the send_key with `wait_screen_change` or just make it explicit with assert_screens which is often the better, more stable and even more time efficient option as the test can progress as soon as it has found the right screen. And multiple `sleep 1` still add up a lot of waiting time which can be saved by assert_screen :)

pevik replied:
I'll try `wait_screen_change`, but I want to avoid requesting users to change needles.

okurz commented on lib/bootloader_setup.pm:
```diff
@@ -56,6 +58,91 @@ sub stop_grub_timeout {
+    $distro .= get_required_var('VERSION');
+    my $cnt_orig = script_output("grep -c 'menuentry .$distro' grub.cfg.old");
+    my $cnt_new  = script_output("grep -c 'menuentry .$distro' /boot/grub2/grub.cfg");
+    die("Unexpected number of grub entries: $cnt_new, expected: " . ($cnt_orig * 2)) if ($cnt_orig * 2 != $cnt_new);
```
You should not rely on `script_output` returning exactly one token. It might include additional output on the serial port which you did not expect. Looking for a regex in the returned output is normally fine but taking the whole output as is and comparing it will not work as you expect

pevik replied:
I did quite lot of testing on different distros and it works well. If this happen, shouldn't it be fixed in `script_output`? I'd really prefer to keep it until any error proven. This can be saved by regexp (read only `^\d+` and ignore the rest or having optional parameter to avoid checks).

okurz replied:
https://progress.opensuse.org/issues/30613 is the corresponding ticket for "can it not be fixed in script_output" and it's 5 months old and so far the answer is: "We don't know how". Seriously, trust me here: You don't want this to fail. But of course you are free to test more on ppc64le and aarch64 where we saw these problems more often :)

okurz commented on lib/bootloader_setup.pm:
```diff
@@ -56,6 +58,91 @@ sub stop_grub_timeout {
+
+And of course the new entries have ima_policy=tcb added to kernel parameters.
+
+=cut
```
IMHO you should use a bit more POD formatting in your documentation, e.g. wrap the parameters in `C<>`. All not wrapped content should be plain english, `GRUP_PARAM` is certainly not :)

pevik replied:
Thanks, I'll fix it.

okurz commented on lib/bootloader_setup.pm:
```diff
@@ -56,6 +58,91 @@ sub stop_grub_timeout {
+And of course the new entries have ima_policy=tcb added to kernel parameters.
+
+=cut
+sub add_custom_grub_entries {
```
The current file lib/bootloader_setup.pm is already more than 700 lines long which is really hard to maintain and also this function does not quite fit what other functions are doing. All the others if I am not mistaken handle the boot menu in place, this one updates grub files within the booted system before rebooting into a new system. I guess we can find a better place

pevik replied:
Well it does bootloader setup, as grub is a boot loader :). And this file is handling grub, so it perfectly fits to it. I'll create `lib/grub_utils.pm` if you want, but IMHO it's confusing. 
And I'd like to avoid doing cleanup with deciding what should be in this new module and what should not. If there is a need of cleanup bootloader setup, it should be done separately and not blocking fixing IMA tests, which has been waiting long time to be fixed.

okurz replied:
Ok, I was not aware that any IMA tests need urgent fixing. Your changes looked more like "Extend test coverage" which should normally be handled with a bit more care to not introduce new problems :) Ok, leave it for the time being :)

pevik commented:
@okurz I've changed everything, but the file where it's located. Do you require to put it somewhere else? If yes, please provide the name. Thx!

okurz commented on tests/kernel/boot_ltp.pm:
```diff
@@ -14,16 +14,24 @@ use 5.018;
-    $self->wait_boot(ready_time => 500);
+    if ($is_ima) {
+        # boot kernel with IMA parameters
+        $self->boot_grub_item(3);
```
This still looks very *magic* to me, whatever "3" means in this context but if you are happy with it then keep it :)

pevik replied:
@okurz 3rd line in grub. It's quite well described at the function itself...

foursixnine replied:
I agree with @okurz. However by looking at the function, looks Ok I guess :) If you're expecting as many menu options :+1:

pevik replied:
@okurz @foursixnine I'm sorry, you're right. ATM `add_custom_grub_entries()` (added in this pull request) doesn't allow to add more than one menu with extra parameter (which might be limitation for the future, I'm not sure how many different grub entries will be needed for IMA tests, which would leed to change in `add_custom_grub_entries()`).
So far is `boot_grub_item()` basically useful for booting this new option. Therefore I can change it to be without parameter and boot this third option, or set `3` as default (if you think this useful to keep selecting it for someone in the future).

foursixnine replied:
I would be ok with a default :)

pevik replied:
OK, I'll change it on Monday. Thanks for comment and review to both.

foursixnine commented on lib/bootloader_setup.pm:
```diff
@@ -56,6 +58,108 @@ sub stop_grub_timeout {
+    $cnt_new = script_output("grep -c 'menuentry .$distro.*($grub_param)' /boot/grub2/grub.cfg");
+    die("Unexpected number of new grub entries: $cnt_new, expected: " . ($cnt_orig)) if ($cnt_orig != $cnt_new);
+    $cnt_new = script_output("grep -c 'linux.*/boot/vmlinu.* $grub_param ' /boot/grub2/grub.cfg");
+    die("Unexpected number of new grub entries with '$grub_param': $cnt_new, expected: " . ($cnt_orig)) if ($cnt_orig != $cnt_new);
```
:+1:
